core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1119,6 +1119,34 @@ public List<SqlNode> fieldList() {
       };
     }
 
+    /**
+     * Checks if a generated ORDER BY node is a literal value.
+     * @param node to check
+     * @return true if node is representing a literal value, false otherwise.
+     */
+    boolean isLiteral(SqlNode node) {
+      // In dialects that use ordinals, NumericLiterals are ordinals and should be kept.
+      if (dialect.getConformance().isSortByOrdinal() && node instanceof SqlNumericLiteral) {
+        return false;
+      }
+      if (node instanceof SqlLiteral) {
+        return true;
+      } else {
+        // The literal value may be wrapped in SqlBasicCalls applying postfix operators.
+        while (node instanceof SqlBasicCall) {
```

Comment:
testing for SqlBasicCall is too broad. `isLiteral('a' || x)` would currently return true, and that's wrong.

I think you're only interested in stripping away DESC, NULLS FIRST, NULLS LAST. Is that true?

If that is true, then you might do better checking for literal in the `RelFieldCollation` (i.e. before you call `toSql(field)`)

Reply:
Such a good call, in testing it looks like postfix operators aren't applied to the node created from field(fieldCollation.getFieldIndex()), so testing for literals directly is possible.